File: DialedInCoffee/RulesEngine/Condition.swift
import Foundation

public enum Condition: Hashable, Sendable {
    case grinderAtLeast(Int)
    case grinderAtMost(Int)
    case strengthAtLeast(Int)
    case strengthAtMost(Int)
    case volumeAtLeast(Int)
    case volumeAtMost(Int)
    case milkSecondsAtLeast(Int)
    case milkSecondsAtMost(Int)
    case tempEquals(TemperatureLevel)
    case drinkIs(DrinkType)
    case drinkIsMilkBased
    case milkKindIs(MilkKind)
    case beanRoastAtLeast(RoastLevel)
    case beanRoastAtMost(RoastLevel)

    public func isSatisfied(
        current: MachineSettings,
        drink: DrinkType,
        milk: Milk,
        bean: BeanSnapshot?
    ) -> Bool {
        switch self {
        case .grinderAtLeast(let n): return current.grinder >= n
        case .grinderAtMost(let n): return current.grinder <= n
        case .strengthAtLeast(let n): return current.strength >= n
        case .strengthAtMost(let n): return current.strength <= n
        case .volumeAtLeast(let n): return current.volumeML >= n
        case .volumeAtMost(let n): return current.volumeML <= n
        case .milkSecondsAtLeast(let n): return current.milkSeconds >= n
        case .milkSecondsAtMost(let n): return current.milkSeconds <= n
        case .tempEquals(let t): return current.temperature == t
        case .drinkIs(let d): return drink == d
        case .drinkIsMilkBased: return drink.isMilkBased
        case .milkKindIs(let k): return milk.kind == k
        case .beanRoastAtLeast(let r):
            guard let bean else { return false }
            return Self.roastOrder(bean.roastLevel) >= Self.roastOrder(r)
        case .beanRoastAtMost(let r):
            guard let bean else { return false }
            return Self.roastOrder(bean.roastLevel) <= Self.roastOrder(r)
        }
    }

    private static func roastOrder(_ r: RoastLevel) -> Int {
        switch r {
        case .light: return 0
        case .mediumLight: return 1
        case .medium: return 2
        case .mediumDark: return 3
        case .dark: return 4
        }
    }
}

File: android/core/src/main/kotlin/com/cawfee/domain/rules/Condition.kt
package com.cawfee.domain.rules

import com.cawfee.domain.model.BeanSnapshot
import com.cawfee.domain.model.DrinkType
import com.cawfee.domain.model.MachineSettings
import com.cawfee.domain.model.Milk
import com.cawfee.domain.model.MilkKind
import com.cawfee.domain.model.RoastLevel
import com.cawfee.domain.model.TemperatureLevel

/** A guard that gates a rule's condition multiplier. Ported from Condition.swift. */
sealed interface Condition {
    data class GrinderAtLeast(val n: Int) : Condition
    data class GrinderAtMost(val n: Int) : Condition
    data class StrengthAtLeast(val n: Int) : Condition
    data class StrengthAtMost(val n: Int) : Condition
    data class VolumeAtLeast(val n: Int) : Condition
    data class VolumeAtMost(val n: Int) : Condition
    data class MilkSecondsAtLeast(val n: Int) : Condition
    data class MilkSecondsAtMost(val n: Int) : Condition
    data class TempEquals(val t: TemperatureLevel) : Condition
    data class DrinkIs(val d: DrinkType) : Condition
    data object DrinkIsMilkBased : Condition
    data class MilkKindIs(val k: MilkKind) : Condition
    data class BeanRoastAtLeast(val r: RoastLevel) : Condition
    data class BeanRoastAtMost(val r: RoastLevel) : Condition

    fun isSatisfied(current: MachineSettings, drink: DrinkType, milk: Milk, bean: BeanSnapshot?): Boolean =
        when (this) {
            is GrinderAtLeast -> current.grinder >= n
            is GrinderAtMost -> current.grinder <= n
            is StrengthAtLeast -> current.strength >= n
            is StrengthAtMost -> current.strength <= n
            is VolumeAtLeast -> current.volumeML >= n
            is VolumeAtMost -> current.volumeML <= n
            is MilkSecondsAtLeast -> current.milkSeconds >= n
            is MilkSecondsAtMost -> current.milkSeconds <= n
            is TempEquals -> current.temperature == t
            is DrinkIs -> drink == d
            DrinkIsMilkBased -> drink.isMilkBased
            is MilkKindIs -> milk.kind == k
            is BeanRoastAtLeast -> bean != null && bean.roastLevel.order >= r.order
            is BeanRoastAtMost -> bean != null && bean.roastLevel.order <= r.order
        }
}
